### DISCUSSION TITLE
Get data from one form and pass the data directly to another function

### DISCUSSION THREAD

**Discussion by vaupunkt** [neuefische/questions#653] [Web] [answered]
I am currently working on this branch
[https://github.com/vaupunkt/cbaspCompanion/tree/feature/kieslerKreisInChatbot
](https://github.com/vaupunkt/cbaspCompanion/tree/feature/kieslerKreisInChatbot
)and this US
[https://github.com/users/vaupunkt/projects/2?pane=issue&itemId=34345612](https://github.com/users/vaupunkt/projects/2?pane=issue&itemId=34345612)

in [`components/ChatBotComponent/index.js`](https://github.com/vaupunkt/cbaspCompanion/blob/feature/kieslerKreisInChatbot/components/ChatBotComponent/index.js)
I am getting data from another component with this function:
```
function handleNext(event) {
    event.preventDefault();
    const formData = new FormData(event.target);
    let dataset = Object.fromEntries(formData);
    handleKieslerkreisInputChange({ ...dataset });
    console.log("Dataset from Kieslerkreis", dataset);
  }
```
the` console.log `is giving me exactly the data I need. with `handleKieslerkreisInputChange({ ...dataset })` I am giving this data to a useState defined in my [`app.js`](https://github.com/vaupunkt/cbaspCompanion/blob/feature/kieslerKreisInChatbot/pages/_app.js). 
I tried to have the useState in different files and also directly in the component without any changes
```
  const [kieslerkreisInput, setKieslerkreisInput] = useState({
    behaviorKieslerkreis: "",
    behaviorChangeKieslerkreis: "",
  });

  function handleKieslerkreisInputChange(input) {
    setKieslerkreisInput({ ...input });
  }
```

back again in my  [`components/ChatBotComponent/index.js`](https://github.com/vaupunkt/cbaspCompanion/blob/feature/kieslerKreisInChatbot/components/ChatBotComponent/index.js) I want to use the data to add it to my main-data I get from my form
```
function submitCallback() {
    console.log("kieslerkreisinput in callback: ", kieslerkreisInput);
    let dataset = cf.getFormData(true);

    dataset = {
      ...dataset,
      ...kieslerkreisInput,
      id: uid(),
      date: date.toISOString().slice(0, 10),
      type: dataset.type[0],
    };
...
    cf.addRobotChatResponse(
      "<strong>Du hast es geschafft 🎉</strong>&&Ich leite dich weiter zu deinem Eintrag.&&Dort kannst du deine Eingaben ansehen, bearbeiten und dich im Kieslerkreis einordnen."
    );
    setTimeout(() => {
      router.push(`/entries/${dataset.id}`);
    }, 5000);
  }
```

The console.log in the `submitCallback-Function` is still the old data and not the entered one from the `handleNext-Function`

I have the feeling, that I am standing right on the edge of the solution but am just blind 😩😅

**Comment by Kheirah** [comment, marked as answer]
Hey there 👋🏼 
I just recently entered the capstone support phase, so I hadn't had the time to familiarize myself with your code too much.
I've tried to understand the issue and implement a fix, but I really don't know if I actually fixed it. You can try it out for yourself and add the following to your `ChatBotComponent`:

```js
const kieslerkreisInputRef = useRef(kieslerkreisInput);
```

My thinking is that your `submitCallback` function may produce an unwanted closure over your state, which makes it stale. If that's the case, a reference could solve the issue, hence the useRef hook. Then further down in your code, add another useEffect hook like this:

```js
useEffect(() => {
  kieslerkreisInputRef.current = kieslerkreisInput;
}, [kieslerkreisInput]);
```

Anytime the state changes, your reference is updated. That's all it does.
Finally, the last change:

```js
function submitCallback() {
    console.log("kieslerkreisinput in callback: ", kieslerkreisInputRef.current);
    let dataset = cf.getFormData(true);

    dataset = {
      ...dataset,
      ...kieslerkreisInputRef.current, //THIS LINE IS CHANGED
      id: uid(),
      date: date.toISOString().slice(0, 10),
      type: dataset.type[0],
    };
    if (Object.entries(dataset).filter((data) => data.comparison === "true")) {
      dataset.comparison = "Erwünschtes Ergebnis erreicht.";
    }
//...
```
Here you are passing the ref's current pointer to the dataset by spreading the state.
Like I said, I couldn't properly follow along with your code and I could be wrong. But do test this and see if it helps.

**Reply by vaupunkt** [reply]
YES! 
thank you so much 🙏🏻
This helped and worked 💯
